src/shogun/features/DotFeatures.cpp

```diff
@@ -361,34 +376,52 @@ SGMatrix<float64_t> CDotFeatures::compute_cov(CDotFeatures* lhs, CDotFeatures* r
 	int32_t dim = dims[0];
 
 	SGMatrix<float64_t> cov(dim, dim);
-	linalg::zero(cov);
-
 	SGVector<float64_t> mean = compute_mean(lhs, rhs);
 
-	for (int i = 0; i < 2; i++)
+	if (copy_data_for_speed)
 	{
-		for (int j = 0; j < nums[i]; j++)
+		SGMatrix<float64_t> centered_data(dim, num);
+		for (int i = 0; i < num; i++)
 		{
-			SGVector<float64_t> v = feats[i]->get_computed_dot_feature_vector(j);
-			linalg::add(v, mean, v, 1.0, -1.0);
-			for (int m = 0; m < v.vlen; m++)
+			SGVector<float64_t> v =
+			    i < nums[0] ? lhs->get_computed_dot_feature_vector(i)
+			                : rhs->get_computed_dot_feature_vector(i - nums[0]);
+
+			centered_data.set_column(i, linalg::add(v, mean, 1.0, -1.0));
+		}
+
+		cov = linalg::matrix_prod(centered_data, centered_data, false, true);
+	}
+	else
+	{
+		linalg::zero(cov);
+		for (int i = 0; i < 2; i++)
+		{
+			for (int j = 0; j < nums[i]; j++)
 			{
-				for (int n = 0; n <= m; n++)
+				SGVector<float64_t> v =
+				    feats[i]->get_computed_dot_feature_vector(j);
+				linalg::add(v, mean, v, 1.0, -1.0);
+				for (int m = 0; m < v.vlen; m++)
 				{
-					(cov.matrix)[m*v.vlen+n] += v.vector[m]*v.vector[n];
+					for (int n = 0; n <= m; n++)
+					{
+						(cov.matrix)[m * v.vlen + n] +=
```

Comment:
This could also be linalg.
Generally, I wonder (offtopic) whether we shouldnt have some Covariance class that has an interface to compute incrementally and in batch version. And then have some kind of iterator access in situations like this one.
Thoughts?